src/browser/wisp_network.js

```diff
@@ -0,0 +1,567 @@
+"use strict";
+
+const DOH_SERVER = "https://dns.google/resolve";
+
+/**
+ * @constructor
+ *
+ * @param {BusConnector} bus
+ * @param {*=} config
+ */
+function WispNetworkAdapter(wisp_url, bus, config)
+{
+    this.last_stream = 1;
+
+    this.connections = {0: {congestion: 0}};
+
+    this.congested_buffer = [];
+    this.send_packet = (data, type, stream_id) => {
+        if(this.connections[stream_id].congestion > 0) {
+            if(type === "DATA") {
+                this.connections[stream_id].congestion--;
+            }
+            this.wispws.send(data);
+        } else {
+            this.connections[stream_id].congested = true;
+            this.congested_buffer.push({data: data, type: type});
+        }
+    };
+
+    this.process_incoming_wisp_frame = (frame) => {
+        const view = new DataView(frame.buffer);
+        const stream_id = view.getUint32(1, true);
+        switch(frame[0]) {
+            case 1: // CONNECT
+                // The server should never send this actually
+                dbg_log("Server sent client-only packet CONNECT", LOG_NET);
+                break;
+            case 2: // DATA
+                if(this.connections[stream_id])
+                    this.connections[stream_id].data_callback(frame.slice(5));
+                else
+                    throw new Error("Got a DATA packet but stream not registered. ID: " + stream_id);
+
+
+                break;
+            case 3: // CONTINUE
+                if(this.connections[stream_id]) {
+                    this.connections[stream_id].congestion = view.getUint32(5, true);
+                }
+
+                if(this.connections[stream_id].congested) {
+                    for(const packet of this.congested_buffer) {
+                        this.send_packet(packet.data, packet.type, stream_id);
+                    }
+                    this.connections[stream_id].congested = false;
+                }
+
+                break;
+            case 4: // CLOSE
+                if(this.connections[stream_id])
+                    this.connections[stream_id].close_callback(view.getUint8(5));
+                delete this.connections[stream_id];
+                break;
+            case 5: // PROTOEXT
+                // Not responding, this is wisp v1 client not wisp v2;
```

Comment:
Perhaps we could add a `dbg_log()` here with message like "Wisp server returned PROTOEXT packet that is not supported by this version of client protocol"? (or catch in `default` case as unknown)

Reply:
It isn't technically unknown so I'm hesitant to mark it as that but if there's no response to a protoext packet the server is supposed to downgrade to wisp V1. I think a message like "got a wisp V2 upgrade request, ignoring" could work